lib/nvector_eigen.h

```diff
@@ -0,0 +1,528 @@
+#ifndef MEPBM_NVECTOR_EIGEN_H
+#define MEPBM_NVECTOR_EIGEN_H
+
+#include "sundials/sundials_nvector.h"
+#include "sundials/sundials_types.h"
+#include "eigen3/Eigen/Sparse"
+#include "eigen3/Eigen/Dense"
+#include <memory>
+#include <cmath>
+
+
+
+/**************************************************************************
+ * Functions whose memory locations are given to the N_Vector Ops structure
+ *************************************************************************/
+
+namespace NVectorOperations
+{
+  /// Returns the length of the vector
+  template <typename VectorType>
+  sunindextype
+  N_VGetLength( N_Vector v)
+  {
+    return (static_cast<VectorType*>(v->content))->size();
+  }
+
+
+
+  /// Creates a new N_Vector with the same ops field as an existing N_Vector. Allocates storage for the vector.
+  template <typename VectorType>
+  N_Vector
+  N_VClone( N_Vector w)
+  {
+    N_Vector v = N_VCloneEmpty(w);
+
+    // create memory for the vector in the new N_Vector
+    // the corresponding delete is called in N_VDestroy()
+    auto cloned = new VectorType(w->ops->nvgetlength(w));
+    v->content = cloned;
+    return v;
+  }
+
+
+
+  /// Creates a new N_Vector with the same ops field as an existing N_Vector. Does not allocate storage for the vector.
+  N_Vector
+  N_VCloneEmpty( N_Vector w)
+  {
+    N_Vector v = N_VNewEmpty();
+
+    // copy pointers to functions from existing vector
+    v->ops = w->ops;
+
+    return v;
+  }
+
+
+
+  /// Destroys the N_Vector and frees allocated memory
+  template<typename VectorType>
+  void
+  N_VDestroy( N_Vector v)
+  {
+    if (v->content != nullptr)
+    {
+      auto *content = static_cast<VectorType *>(v->content);
+      delete content;
+      v->content = nullptr;
+    }
+    N_VFreeEmpty(v);
+  }
+
+
+
+  /// Returns storage requirements for one N_Vector. This is a dummy function because it is not of interest.
+  void
+  N_VSpace(N_Vector v, sunindextype* lrw, sunindextype* liw)
+  {}
+
+
+
+  /// Returns a pointer to an array from the N_Vector.
+  template<typename VectorType>
+  realtype*
+  N_VGetArrayPointer(N_Vector v)
+  {
+    auto vector = static_cast<VectorType*>(v->content);
+    return vector->data();
+  }
+
+
+
+  /// Overwrites the pointer to the data in an N_Vector.
+  // FIXME: I'm having trouble with this but I don't think it's necessary so maybe just skip it
```

Comment:
if you don't know what this function should do, do the same `assert(false)` thing as above to make sure the function is never actually called

Reply:
From what I can tell this function is only used when the Jacobian is approximated numerically, so I should never use it in my implementation. So I used the `assert(false)` thing.